rfc-172-graphql-for-govuk.md

```diff
@@ -0,0 +1,509 @@
+|              ->  Index Scan using index_link_sets_on_content_id on link_sets  (cost=0.42..33.77 rows=4 width=40) (actual time=0.012..0.024 rows=4 loops=1)                                                         |
+|                    Index Cond: (content_id = ANY ('{9bdb6017-48c9-4590-b795-3c19d5e59320,14aa298f-03a8-4e76-96de-483efa3d001f,8f8d90e7-5289-4942-82f3-ed03b1a0ac10,cac74185-649e-4e2b-b5b8-2c5ec86c6609}'::uuid[]))|
+|              ->  Index Scan using index_links_on_link_set_id on links  (cost=0.44..220.39 rows=58 width=67) (actual time=0.004..0.006 rows=7 loops=4)                                                              |
+|                    Index Cond: (link_set_id = link_sets.id)                                                                                                                                                        |
+|        ->  Index Scan using index_documents_on_content_id_and_locale on documents document  (cost=0.43..6.13 rows=1 width=47) (actual time=0.003..0.003 rows=1 loops=28)                                           |
+|              Index Cond: ((content_id = links.target_content_id) AND ((locale)::text = 'en'::text))                                                                                                                |
+|  ->  Index Scan using index_editions_on_document_id_and_state on editions  (cost=0.43..22.29 rows=7 width=769) (actual time=0.003..0.003 rows=1 loops=28)                                                          |
+|        Index Cond: ((document_id = document.id) AND ((state)::text = 'published'::text))                                                                                                                           |
+|Planning Time: 1.115 ms                                                                                                                                                                                             |
+|Execution Time: 0.318 ms                                                                                                                                                                                            |
++--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+```
+
+Which can be [explored visually in PG Explain][pg-explain-query-plan]
+
+![PG Explain Visualizer showing the execution plan above](rfc-172/pgexplain-query-plan-visualisation.png)
+
+</details>
+
+Still, even with dataloader batching queries and individual queries making good use of indexes, we're asking the database to do at least around 10x more work than the current content-store database has to do, and that will have some performance or infrastructure cost implications. We will need to carefully consider how this solution performs at scale to evaluate these.
```

Comment:
Our database instances are very underutilised right now, so I expect any cost implications would likely be below the noise floor, so to speak.

It's also important to set this valid concern in the wider context of enabling the rationalisation of GOV.UK's database architecture, which has the potential to reduce infrastructure costs and improve performance.

In other words, if the proposal is well executed and the mitigations you suggest are implemented, perceived performance from the end-user perspective is likely to improve while somewhat reducing running costs in the medium to long term.